include/ginkgo/core/base/device.hpp

```diff
@@ -0,0 +1,140 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_DEVICE_HPP_
+#define GKO_PUBLIC_CORE_BASE_DEVICE_HPP_
+
+
+#include <array>
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+
+#include <ginkgo/config.hpp>
+
+
+namespace gko {
+
+
+class CudaExecutor;
+
+class HipExecutor;
+
+
+/**
+ * NvidiaDevice handles the number of executor on Nvidia devices and have the
+ * corresponding recursive_mutex.
+ */
+class NvidiaDevice {
+    friend class CudaExecutor;
+// If Hip is compiled for NVCC, give NvidiaDevice's permission to HipExecutor
+#if GINKGO_HIP_PLATFORM_NVCC
+    friend class HipExecutor;
+#endif
+
+private:
+    /**
+     * get_mutex gets the static mutex reference at i.
+     *
+     * @param i  index of mutex
+     *
+     * @return recursive_mutex reference
+     */
+    static std::recursive_mutex &get_mutex(int i)
+    {
+        static std::recursive_mutex mutex[max_devices];
```

Comment:
Why recursive? Looking at the code here, I don't see a direct recursive requirement (thankfully).  I would rather we use normal mutexes so that things don't behave weirdly if recursion happens at some point.

Reply:
It is required in destructor of executor.
We need one mutex to make sure the delete exec and get_num_execs are thread safe. get_num_exec also needs mutex when calling it from other place. Thus, it needs to be recursive or two mutex.

Reply:
I am also fine with two kind mutex, one for desstructor and other for counter.

Reply:
notes after discussion with @tcojean :
We always use create function to create the shared_ptr of Executor, so the destruct behavior can all be done when we create the `shared_ptr`.
In the shared_ptr deleter, we use a mutex and directly work on the Device not the Executor decrease/get num_execs.
Thus, we can still use the `std::mutex` not `std::recursive_mutex`